core/src/main/java/com/google/zxing/common/ECIStringBuilder.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright 2022 ZXing authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.zxing.common;
+
+import com.google.zxing.FormatException;
+
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Class that converts a sequence of ECIs and bytes into a string
+ *
+ * @author Alex Geller
+ */
+public final class ECIStringBuilder {
+  private StringBuilder currentBytes;
+  private StringBuilder result;
+  private Charset currentCharset = StandardCharsets.ISO_8859_1;
+
+  public ECIStringBuilder() {
+    currentBytes = new StringBuilder();
+  }
+  public ECIStringBuilder(int initialCapacity) {
+    currentBytes = new StringBuilder(initialCapacity);
+  }
+
+  /**
+   * Appends {@code value} as a byte value
+   */
+  public void append(char value) {
+    currentBytes.append((char) (value & 0xff));
+  }
+
+  /**
+   * Appends {@code value} as a byte value (not its string representation)
+   */
+  public void append(byte value) {
+    currentBytes.append((char) (value & 0xff));
+  }
+
+  /**
+   * Appends the characters in {@code value} as bytes values
+   */
+  public void append(String value) {
+    currentBytes.append(value);
+  }
+
+  /**
+   * Append the string repesentation of {@code value} (short for {@code append(String.valueOf(value))})
+   */
+  public void append(int value) {
+    append(String.valueOf(value));
+  }
+
+  public void appendECI(int value) throws FormatException {
+    encodeCurrentBytesIfAny();
+    CharacterSetECI characterSetECI = CharacterSetECI.getCharacterSetECIByValue(value);
+    if (characterSetECI == null) {
+      throw FormatException.getFormatInstance();
+    }
+    currentCharset = characterSetECI.getCharset();
+  }
+
+  private void encodeCurrentBytesIfAny() {
+    if (currentCharset.equals(StandardCharsets.ISO_8859_1)) {
+      if (currentBytes.length() > 0) {
+        if (result == null) {
+          result = currentBytes;
+          currentBytes = new StringBuilder();
+        } else {
+          result.append(currentBytes);
+          currentBytes = new StringBuilder();
+        }
+      }
+    } else if (currentBytes.length() > 0) {
+      byte[] bytes = currentBytes.toString().getBytes(StandardCharsets.ISO_8859_1);
```

Comment:
Hm but there is a conversion back and forth here and line 93. Why can this not stay in strings?  I imagine there is indeed a good reason in the end, just checking my basic understanding

Reply:
Yes, we get to line 93 if the test in line 79 is negative (testing and finding that the current encoding isn't ISO-8859-1). In this case we have to do a character-byte-character conversion. This happens when appendECI() is called with a non ISO-8859-1 value. In the general case this is not called and also in the case that there is a switchback to ISO-8859-1, we use the efficient path again.